backend/FwLite/MiniLcm/SyncHelpers/EntrySync.cs

```diff
@@ -0,0 +1,124 @@
+
+namespace MiniLcm.SyncHelpers;
+
+public static class EntrySync
+{
+    public static async Task<int> Sync(Entry[] afterEntries,
+        Entry[] beforeEntries,
+        IMiniLcmApi api)
+    {
+        Func<IMiniLcmApi, Entry, Task<int>> add = static async (api, afterEntry) =>
+        {
+            await api.CreateEntry(afterEntry);
+            return 1;
+        };
+        Func<IMiniLcmApi, Entry, Task<int>> remove = static async (api, beforeEntry) =>
+        {
+            await api.DeleteEntry(beforeEntry.Id);
+            return 1;
+        };
+        Func<IMiniLcmApi, Entry, Entry, Task<int>> replace = static async (api, beforeEntry, afterEntry) => await Sync(afterEntry, beforeEntry, api);
+        return await DiffCollection.Diff(api, beforeEntries, afterEntries, add, remove, replace);
+    }
+
+    public static async Task<int> Sync(Entry afterEntry, Entry beforeEntry, IMiniLcmApi api)
+    {
+        var updateObjectInput = EntryDiffToUpdate(beforeEntry, afterEntry);
+        if (updateObjectInput is not null) await api.UpdateEntry(afterEntry.Id, updateObjectInput);
+        var changes = await SensesSync(afterEntry.Id, afterEntry.Senses, beforeEntry.Senses, api);
+
+        changes += await Sync(afterEntry.Components, beforeEntry.Components, api);
+        changes += await Sync(afterEntry.ComplexForms, beforeEntry.ComplexForms, api);
+        changes += await Sync(afterEntry.Id, afterEntry.ComplexFormTypes, beforeEntry.ComplexFormTypes, api);
+        return changes + (updateObjectInput is null ? 0 : 1);
+    }
+
+    private static async Task<int> Sync(Guid entryId,
+        IList<ComplexFormType> afterComplexFormTypes,
+        IList<ComplexFormType> beforeComplexFormTypes,
+        IMiniLcmApi api)
+    {
+        return await DiffCollection.Diff(api,
+            beforeComplexFormTypes,
+            afterComplexFormTypes,
+            complexFormType => complexFormType.Id,
+            async (api, afterComplexFormType) =>
+            {
+                await api.AddComplexFormType(entryId, afterComplexFormType.Id);
+                return 1;
+            },
+            async (api, beforeComplexFormType) =>
+            {
+                await api.RemoveComplexFormType(entryId, beforeComplexFormType.Id);
+                return 1;
+            },
+            //do nothing, complex form types are not editable, ignore any changes to them here
+            static (api, beforeComplexFormType, afterComplexFormType) => Task.FromResult(0));
+    }
+
+    private static async Task<int> Sync(IList<ComplexFormComponent> afterComponents, IList<ComplexFormComponent> beforeComponents, IMiniLcmApi api)
+    {
+        return await DiffCollection.Diff(api,
+            beforeComponents,
+            afterComponents,
+            //we can't use the ID as there's none defined by Fw so it won't work as a sync key
+            component => (component.ComplexFormEntryId, component.ComponentEntryId, component.ComponentSenseId),
```

Comment:
So, this will never result in a `replace` operation, because the identity is equivalent to the value, so the collection differ will call `replace` for all unchanged components, bceause the id is the same and then the replace logic starting on L83 will do nothing, because the value is the same.

You probably know that and it's the simplest way to keep everything happy for now.

This is where a more sophisticated list differ would be nice in the future. E.g. if the before and after lists are the same length and only the nth object changed, then a replace probably makes sense even if the IDs are different. But then we'd be disregarding Guids that the client generated, so that might be bad.

Reply:
yeah I came to the same conclusion yesterday and decided to not do anything about it. Sad since I did all that work on tests and such for all those different cases but yeah I just don't see how a replace would make sense, especially considering there is no Id from Fieldworks and this code needs to handle both, I'd rather them be consistent.

Reply:
I ended up actually removing all this and throwing if it gets reached for some reason